### ISSUE TITLE
Add buffer()-compatible function for Python 3 to twisted.python.compat

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6099]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6099|
|Type|enhancement|
|Created|2012-10-11 18:49:21Z|
|Branch|https://github.com/twisted/twisted/tree/buffer-py3-6099|

`buffer()` is used in `twisted.internet.tcp`, `twisted.internet.abstract` (and elsewhere: IOCP and `unix.py`). We should provide a sufficiently compatible version based on memoryview to meet the requirements of those users.

<details><summary>Searchable metadata</summary>

```
trac-id__6099 6099
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_buffer_py3_6099 branches/buffer-py3-6099
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1349981361000000 1349981361000000
changetime__1350073801000000 1350073801000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

But probably not based on `memoryview`, due to the problems it comes with (complexity, poor documentation, incompatibility with parts of pyOpenSSL, etc).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

(In [36118]) Branching to 'buffer-py3-6099'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst set owner to @exarkun|
|-|-|

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/buffer-py3-6099 is running.

Note that the argument names were based on the Python 2 `buffer()` argument names - http://docs.python.org/library/functions.html#buffer

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @itamarst|
|-|-|

Regarding naming... How about a name that doesn't give the appearance of being more than trivially `buffer`-compatible - `lazyByteSlice` or something descriptive like that?  The idea behind this being that our requirement is not for a `buffer` replacement (and this isn't one, since that would be a ton of work), but for a way to slice strings without copying memory unnecessarily.  `buffer` did that for us, but is kind of an implementation detail.

Alternatively (and I think this is worse), be more explicit in documenting the limitations of our `buffer`.

Otherwise this looks great.  Please address the naming point somehow and then merge.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst set status to `closed`|
|-|-|

(In [36128]) Merge buffer-py3-6099.

Author: itamar
Review: exarkun
Fixes: [#6099](https://github.com/twisted/twisted/issues/6099)

Add function to emulate buffer() on both Python 2 and (less efficiently) on Python 3.
